### PR TITLE
[CS-4126] Prompt for permissions on Pay Tab has white overlay on Android

### PR DESCRIPTION
**Pull request by oidoug**
### Description

The old camera module we were using handles camera permission while trying to render, with the new rn-nav6 the screen was getting blank while permission was requested, and it is not being maintained anymore. 

This PR upgrades our camera module to the better-maintained expo-camera and expo-barcode-scanner (both packages are needed for QR reading).

- [x] Completes #(CS-4126)

* known issue: Camera blacks out after navigation "go backs" even though it should be rendered, something related to Camera only allowing one instance at a time. * - I'll handle this in another PR.

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS
- [x] Tested on Android

### Screenshots

https://user-images.githubusercontent.com/129619/175535292-005471ec-3278-4c5d-a81c-2b00236cc490.mp4

### CODE REVIEW

File: cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/useQRCodeScannerPage.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+export const useQRCodeScannerPage = (props: useScannerParams) => {
+  const [isCameraAllowed, setCameraAllowed] = useBooleanState();
+  const { onScan, isLoading } = useScanner(props);
+  const { result: isEmulator } = useIsEmulator();
+  const [error, showError] = useBooleanState();
+  const { network } = useAccountSettings();
+  const isFocused = useIsFocused();
+
+  const requestPersmission = useCallback(async () => {
```
Comment (danibonilha) [inline cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/useQRCodeScannerPage.tsx]: typo `requestPersmission` -> `requestPermission`

File: cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/components/CameraNotAuthorizedView.tsx [hunk 2]
```diff
@@ -34,7 +33,7 @@ const CameraNotAuthorizedView = () => {
       >
         {strings.authorize.info}
       </Text>
-      <Button onPress={handlePressSettings}>
+      <Button onPress={enableCameraPressed}>
```
Comment (danibonilha) [inline cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/components/CameraNotAuthorizedView.tsx]: nit: we could follow the pattern, like, `onEnableCameraPress`

File: cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/useQRCodeScannerPage.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      return;
+    }
+
+    requestPersmission();
+  }, [requestPersmission]);
+
+  useEffect(() => {
+    if (isFocused) requestPersmission();
+  }, [requestPersmission, isFocused]);
+
```
Comment (danibonilha) [inline cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/useQRCodeScannerPage.tsx]: Shouldn't we check if it is not authorized also  ?  `  if (isFocused && !isCameraAllowed)`

File: src/useAppInit.ts [hunk 4]
```diff
@@ -97,9 +101,29 @@ export const useAppInit = () => {
...
+      unauthorizeDelay.current = null;
+      return;
     }
+
+    // We'll only delay the unauthorization on android and in these specific screens.
+    if (
+      Device.isAndroid &&
+      (Navigation.getActiveRouteName() === Routes.QR_SCANNER_SCREEN ||
+        Navigation.getActiveRouteName() === Routes.PAY_MERCHANT)
+    ) {
```
Comment (danibonilha) [inline src/useAppInit.ts]: why we need to check for PayMerchant ? if we do need both, there's some nit code to improve this
```tsx
const tempAuthorizedRoutes: string[] = [Routes.QR_SCANNER_SCREEN, Routes.PAY_MERCHANT]

const currentRoute = Navigation.getActiveRouteName() || ''
const isOnDelayedRoute = tempAuthorizedRoutes.includes(currentRoute)

if(Device.isAndroid && isOnDelayedRoute)
```

File: cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/components/CameraNotAuthorizedView.tsx (hunk 2 above)
Comment (oidoug) [reply cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/components/CameraNotAuthorizedView.tsx]: yes, that's better.

File: src/useAppInit.ts [hunk 5]
```diff
@@ -45,6 +48,7 @@ export const useAppInit = () => {
   useNotificationSetup();
 
   const initialDeepLink = useRef<string | null>(null);
+  const unauthorizeDelay = useRef<number | null>(null);
```
Comment (danibonilha) [inline src/useAppInit.ts]: nite: Not sure naming this as a delay makes sense, maybe this should be called `androidLastAuthorizedDate` and isn't the type `useRef<Date | null>`

File: src/useAppInit.ts [hunk 6]
```diff
@@ -97,9 +101,29 @@ export const useAppInit = () => {
   }, [isAuthorized, walletReady]);
 
   useEffect(() => {
-    if (movedToBackground) {
-      setUserUnauthorized();
+    // Resuming state, we check if we should unauthorize already.
+    if (!movedToBackground && unauthorizeDelay.current) {
```
Comment (danibonilha) [inline src/useAppInit.ts]: not quite sure I understand this, checking against the movedBackground, if we want to check resuming the state, wouldn't justBecameActive work for this case ?

File: src/useAppInit.ts [hunk 7]
```diff
@@ -97,9 +101,29 @@ export const useAppInit = () => {
...
+        Navigation.getActiveRouteName() === Routes.PAY_MERCHANT)
+    ) {
+      unauthorizeDelay.current = Date.now();
+
+      return;
+    }
+
+    // Otherwise, unauthorize right away.
+    setUserUnauthorized();
   }, [setUserUnauthorized, movedToBackground]);
```
Comment (danibonilha) [inline src/useAppInit.ts]: lmk what you think about this `useEffect`, if it works for this case 
```ts
  useEffect(() => {
    if (movedToBackground) {
      const tempAuthorizedRoutes: string[] = [
        Routes.QR_SCANNER_SCREEN,
        Routes.PAY_MERCHANT,
      ];

      const isOnDelayedRoute = tempAuthorizedRoutes.includes(
        Navigation.getActiveRouteName() || ''
      );

      if (Device.isAndroid && isOnDelayedRoute) {
        const currentDate = Date.now();

        if (!unauthorizedDelay.current) {
          unauthorizedDelay.current = currentDate;
          return;
        }

        if (currentDate - unauthorizedDelay.current < UNAUTHORIZE_DELAY) {
          return;
        }
      }

      unauthorizedDelay.current = null;

      setUserUnauthorized();
    }
  }, [setUserUnauthorized, movedToBackground]);
```

File: cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/useQRCodeScannerPage.tsx (hunk 3 above)
Comment (oidoug) [reply cardstack/src/screens/QRScannerScreen/pages/QRCodeScanner/useQRCodeScannerPage.tsx]: Yes, that makes sense.

File: src/useAppInit.ts (hunk 4 above)
Comment (oidoug) [reply src/useAppInit.ts]: We do, once QR is read and we navigate to prepaid card selection, the app jumps to background again.

We could try finding out why that happens, but might be a bit more challenging.

File: src/useAppInit.ts (hunk 5 above)
Comment (oidoug) [reply src/useAppInit.ts]: `Date.now()` return the milliseconds since epoch.

File: src/useAppInit.ts (hunk 6 above)
Comment (oidoug) [reply src/useAppInit.ts]: `movedToBackground` is false when app becomes active again, I think it's safe to use it.

File: src/useAppInit.ts (hunk 7 above)
Comment (oidoug) [reply src/useAppInit.ts]: Hum, almost, but we need to check unauthorizedDelay when app gets active again (!movedToBackgroun) or we'll only be able to unauthorize when going to background again.

I see my algo is a bit confusing, I'll try merge both.

**Comment by oidoug** [comment]
As suggested, I'm splitting android's background logout delay from camera lib upgrade PR.

**Comment by oidoug** [comment]
Upgrading RN to latest fixed the white overlay this PR was addressing.
expo-camera is not working as well as our current module, so this PR is being dropped.

**Pull request closed without merging**